### PR TITLE
    Automatic pooling merge

### PR DESCRIPTION
**Pull request by natsala13**
* Still need to squash all commits to one.
 * Comments exists also on docs file.
 * Late changes merged to this branch - I think there are still adaptation to make.

### CODE REVIEW

File: generate_encoder.py [hunk 1]
```diff
@@ -246,28 +266,30 @@ def motion_fusion(args, device, edge_rot_dict_general, encoder, eval_ids, g_ema,
...
+
+
+def manual_editing(encoder, static: StaticData, mean_joints: torch.Tensor, std_joints: torch.Tensor,
+                    g_ema, output_path_anim, make_mask, eval_ids, motion_data, use_velocity, device, **kwargs):
     output_path_anim_id = osp.join(output_path_anim, f'spatial_editing')
     os.makedirs(output_path_anim_id, exist_ok=True)
     for eval_id in tqdm(eval_ids):
         motion_data_eval = torch.from_numpy(motion_data[[eval_id]]).float().to(device)
-        save_bvh = functools.partial(motion2bvh, edge_rot_dict_general=edge_rot_dict_general, entity='Edge')
+        # save_bvh = functools.partial(motion2bvh, edge_rot_dict_general=edge_rot_dict_general, entity='Edge')
```
Comment (Inbal-leibovitch) [inline generate_encoder.py]: delete commented line

File: generate_encoder.py [hunk 2]
```diff
@@ -246,28 +266,30 @@ def motion_fusion(args, device, edge_rot_dict_general, encoder, eval_ids, g_ema,
...
         joint_idx = 12
         order ='xyz'
         for axis in [1]:
-            motion_data_edit = to_list_4D(motion_data_eval.detach().cpu().numpy())
-            motion_data_edit = un_normalize(motion_data_edit, mean=edge_rot_dict_general['mean'].transpose(0, 2, 1, 3),
-                                       std=edge_rot_dict_general['std'].transpose(0, 2, 1, 3))
+            # motion_data_edit = to_list_4D(motion_data_eval.detach().cpu().numpy())
```
Comment (Inbal-leibovitch) [inline generate_encoder.py]: delete commented line

File: models/encoder_mask.py [hunk 3]
```diff
@@ -114,28 +117,29 @@ def __call__(self, pred, target):
...
+        # edge_rot_dict_general = self.edge_rot_dict_general
+        motion_data = motion_data * self.std_joints[:, :, :motion_data.shape[2]] + \
+                      self.mean_joints[:, :, :motion_data.shape[2]]
         motion_for_fk = motion_data.transpose(1,
                                               3)  # samples x features x joints x frames  ==>  samples x frames x joints x features
         if self.use_glob_pos:
             #  last 'joint' is global position. use only first 3 features out of it.
             glob_pos = motion_for_fk[:, :, self.pos_offset, :3]
-            if 'use_velocity' in edge_rot_dict_general and edge_rot_dict_general['use_velocity']:
+            # if 'use_velocity' in edge_rot_dict_general and edge_rot_dict_general['use_velocity']:
```
Comment (Inbal-leibovitch) [inline models/encoder_mask.py]: delete commented line

File: models/encoder_mask.py [hunk 4]
```diff
@@ -149,29 +153,30 @@ def __call__(self, pred, target):
...
-                            edge_rot_dict_general['offset_root'],
-                            axis=0)
-        offsets = torch.from_numpy(offsets).to(device).type(torch.float32)
+    def __init__(self, static, device, use_glob_pos, use_contact, use_velocity,
+                 mean_joints, std_joints, local_frame=False):
         self.use_glob_pos = use_glob_pos
-        self.fk = ForwardKinematicsJoint(edge_rot_dict_general['parents_with_root'], offsets)
-        self.edge_rot_dict_general = edge_rot_dict_general
+        self.fk = ForwardKinematicsJoint(static.parents, static.offsets)
+        # self.edge_rot_dict_general = edge_rot_dict_general
```
Comment (Inbal-leibovitch) [inline models/encoder_mask.py]: delete commented line

File: models/encoder_mask.py [hunk 5]
```diff
@@ -149,29 +153,30 @@ def __call__(self, pred, target):
...
+        self.use_velocity = use_velocity
+
+        self.mean_joints = mean_joints
+        self.std_joints = std_joints
 
     def get_pos(self, motion_data):
-        edge_rot_dict_general = self.edge_rot_dict_general
-        motion_data = motion_data * edge_rot_dict_general['std_tensor'][:, :, :motion_data.shape[2]] + \
-                      edge_rot_dict_general['mean_tensor'][:, :, :motion_data.shape[2]]
+        # edge_rot_dict_general = self.edge_rot_dict_general
```
Comment (Inbal-leibovitch) [inline models/encoder_mask.py]: delete

File: evaluate.py [hunk 6]
```diff
@@ -55,46 +56,51 @@ def generate(args, g_ema, device, mean_joints, std_joints, entity):
         index = range(len(generated_motion))
 
     generated_motion_np, _ = get_gen_mot_np(args, generated_motion, mean_joints, std_joints)
-    generated_motions = np.concatenate(generated_motion_np, axis=0)
 
-    if entity.str() == 'Joint':
```
Comment (Inbal-leibovitch) [inline evaluate.py:60]: what happens if entity is Joint? like in the model trained on Humanact12

File: train.py [hunk 7]
```diff
@@ -508,24 +546,24 @@ def main(args_not_parsed):
...
+    #
+    # fig = motion2fig(motion_data, H=512, W=512, entity=args.entity,
+    #                  edge_rot_dict_general=edge_rot_dict_general)
+    # fig_name = osp.join(images_output_folder, 'real_motion.png')
+    # fig.savefig(fig_name, dpi=300, bbox_inches='tight')
+    # plt.close() # close figure
+    # ###################################################
+    #
+    # if args.clearml:
+    #     logger.report_media(title='Image', series='Ground Truth Motion', iteration=0, local_path=fig_name)
```
Comment (Inbal-leibovitch) [inline train.py]: why is all of this commented out?

File: train_encoder.py [hunk 8]
```diff
@@ -346,10 +342,12 @@ def main(args_not_parsed):
         e_optim.load_state_dict(ckpt["e_optim"])
 
     gt_bone_lengths = calc_bone_lengths(motion_data) if args.entity == 'Joint' else None
+    dynamic_motion0 = DynamicData(motion_data[0], static, use_velocity=args.use_velocity)
 
     motion_path = osp.join(animations_output_folder, 'real_motion.bvh')
-    motion2bvh(motion_data[0], motion_path, parents=entity.parents_list, entity=args.entity,
-               edge_rot_dict_general=edge_rot_dict_general)
+    motion2bvh_rot(dynamic_motion0, motion_path)
+    # motion2bvh_rot(motion_data[0], motion_path, parents=entity.parents_list, entity=args.entity,
```
Comment (Inbal-leibovitch) [inline train_encoder.py]: delete comment

File: utils/foot.py [hunk 9]
```diff
@@ -1,46 +1,62 @@
 import torch
-from models.kinematics import ForwardKinematicsJoint
 import numpy as np
+from models.kinematics import ForwardKinematicsJoint
 
+# from utils.data import foot_names
+# from motion_class import StaticData  # Cycle import
```
Comment (Inbal-leibovitch) [inline utils/foot.py]: delete comments

File: utils/foot.py [hunk 10]
```diff
@@ -1,46 +1,62 @@
...
-    names = edge_rot_dict_general['names_with_root']
-    n_foot = len(foot_names)
-    idx_foot = [np.nonzero(names == foot_name)[0][0] for foot_name in foot_names]
-    foot_location = joint_location[:, :, idx_foot]
+
+    foot_indexes = static.foot_indexes[-1]
+    foot_location = joint_location[:, :, foot_indexes]
+
+    # up_joints = joint_location[..., 1]
+    # res = torch.sort(up_joints.reshape(-1))[0]
```
Comment (Inbal-leibovitch) [inline utils/foot.py]: delete

File: utils/visualization.py [hunk 11]
```diff
@@ -117,68 +133,40 @@ def motion2fig_1_motion_3_angles(data, H=512, W=512):
...
-    n_frames = data.shape[-1]
-    if n_sampled_frames is None:
-        n_sampled_frames = n_frames # how many samples to take from the whole video
-    sampled_frames = np.linspace(0, n_frames-1, n_sampled_frames).round().astype(int)
+def motion2fig(static: StaticData, dynamics: DynamicData,
+               height=512, width=512, n_sampled_motions=5, n_sampled_frames=5):
+    # n_sampled_motions = min(n_sampled_motions, dynamics.n_frames, 10)
+    # sampled_frames = np.linspace(0, dynamics.n_frames-1, n_sampled_frames).round().astype(int)
+    # print(sampled_frames)  # [ 0 16 32 47 63]
+    # dynamics = dynamics.sample_frames(sampled_frames)
```
Comment (Inbal-leibovitch) [inline utils/visualization.py]: remove commented lines

File: evaluate.py (hunk 6 above)
Comment (sigal-raab) [reply evaluate.py:60]: Natan, following Inbal's comment, you can apply this evaluation script on the humanact12 trained model (see one of the latest issues to find it). Also, apply on a MoDi trained model. For both cases, make sure the paper's metrics are reconstructed.

File: evaluate.py [hunk 12]
```diff
@@ -55,46 +56,51 @@ def generate(args, g_ema, device, mean_joints, std_joints, entity):
...
+    return generated_motions
 
-def convert_motions_to_location(args, generated_motion_np, edge_rot_dict_general):
-    edge_rot_dict_general['std_tensor'] = edge_rot_dict_general['std_tensor'].cpu()
-    edge_rot_dict_general['mean_tensor'] = edge_rot_dict_general['mean_tensor'].cpu()
-    if args.dataset == 'mixamo':
-        edge_rot_dict_general['offsets_no_root'] /= 100 ## not needed in humanact
 
+def convert_motions_to_location(dynamics, dataset_type):
     generated_motions = []
```
Comment (sigal-raab) [inline evaluate.py:88]: Throughout the whole code: if not specifically dealing with static/dynamic attributes, don't call a motion 'dynamics'. Call it 'motion', or 'dynamic_motion', or similar.
Do it for all your changes

Comment (sigal-raab) [reply evaluate.py:88]: Here, for example, it is important for the user to know that the argument is a motion and that it's a generated one. So the name 'generated_motion' makes sense. When the user sees the name 'dynamic' he thinks: what is this argument? a dynamic what?

File: generate.py [hunk 13]
```diff
@@ -35,18 +37,20 @@ def interpolate(args, g_ema, device, mean_latent):
...
             W_to = mean_latent
 
-        generated_motion =[None] * num_interp
+        generated_motion = [None] * num_interp
         steps = torch.linspace(0, 1, num_interp, device=device)
 
         for interp_idx in np.arange(num_interp):
             cur_W = W_from.lerp(W_to, steps[interp_idx])
-            generated_motion[interp_idx], _, _ = g_ema(
+            interpolated_motion, _, _ = g_ema(
```
Comment (sigal-raab) [inline generate.py:45]: Are you sure you want to remove interp_idx? Have you tested the interp case?

File: models/encoder_mask.py [hunk 14]
```diff
@@ -114,28 +117,29 @@ def __call__(self, pred, target):
...
```
Comment (sigal-raab) [inline models/encoder_mask.py]: erase

File: motion_class.py [hunk 15]
```diff
@@ -0,0 +1,689 @@
...
+
+PARENTS LIST
+[ ] neighbors_by_distance()
+[ ] len(parents) in traits.py
+[ ] enable_global_position() and enable_foot_contact()
+
+NOTES
+* anim_from_edge_rot_dict apply extend joints on motion
+* expand joints is important otherwise we see some inconsistency at plotting
+
```
Comment (sigal-raab) [inline motion_class.py]: delete all personal notes

File: train.py [hunk 16]
```diff
@@ -22,8 +22,10 @@
 
 from models.gan import Generator, Discriminator
 from utils.foot import get_foot_contact, get_foot_velo
-from utils.data import Joint, Edge # to be used in 'eval'
+from utils.data import Joint # TODO: Whats that? - to be used in 'eval'
```
Comment (sigal-raab) [inline train.py]: delete comment

File: train.py [hunk 17]
```diff
@@ -182,19 +191,12 @@ def train(args, loader, generator, discriminator, g_optim, d_optim, g_ema, devic
...
     time_measure = []
     start_time_measure = time.time()
-    for idx in pbar:
-        i = idx + args.start_iter
-
-        if i > args.iter:
-            print("Done!")
-            break
-
+    for i in pbar:
```
Comment (sigal-raab) [inline train.py:200]: why did you delete the lines that deal with start_iter? they enable utilizing an existing model

File: train.py [hunk 18]
```diff
@@ -508,24 +546,24 @@ def main(args_not_parsed):
...
-    fig = motion2fig(motion_data, H=512, W=512, entity=args.entity,
-                     edge_rot_dict_general=edge_rot_dict_general)
-    fig_name = osp.join(images_output_folder, 'real_motion.png')
-    fig.savefig(fig_name, dpi=300, bbox_inches='tight')
-    plt.close() # close figure
-    if args.clearml:
-        logger.report_media(title='Image', series='Ground Truth Motion', iteration=0, local_path=fig_name)
+    motion_data, mean_joints, std_joints, = motion_from_raw(args, motion_data_raw, static)
+
+    # Just save some real motion for start - TODO: Why?
```
Comment (sigal-raab) [inline train.py:551]: We would like to see what our GT data looks like. Please restore this commented block.

File: train.py (hunk 7 above)
Comment (sigal-raab) [reply train.py]: We would like to see what our GT data looks like. Please restore this commented block.

File: utils/fbx2bvh.py [hunk 19]
```diff
@@ -7,9 +7,10 @@
 
 import bpy
 
+#'/Users/nathansala/tau/code/MoDi/dataset/mau'
 in_dir = '<fbx folder>'  #  example: '/home/username/data/mixamo/fbx/binaries_30_fps'
 out_dir = in_dir.replace('fbx', 'bvh')
-fbx_files = glob(osp.join(in_dir, '*', '*.fbx'))
+fbx_files = os.listdir(out_dir)  # glob(osp.join(in_dir, '*', '*.fbx'))
```
Comment (sigal-raab) [inline utils/fbx2bvh.py]: looks like a bug. out_dir does not contain files yet yet.

File: utils/fbx2bvh.py [hunk 20]
```diff
@@ -7,9 +7,10 @@
 
 import bpy
 
+#'/Users/nathansala/tau/code/MoDi/dataset/mau'
```
Comment (sigal-raab) [inline utils/fbx2bvh.py]: delete

File: utils/fbx2bvh.py [hunk 21]
```diff
@@ -25,13 +26,13 @@
 
     os.makedirs(osp.join(out_dir, rel_dir), exist_ok=True)
 
-    bpy.ops.import_scene.fbx(filepath=in_file)
+    bpy.ops.import_scene.fbx(filepath=osp.join(out_dir, in_file))
```
Comment (sigal-raab) [inline utils/fbx2bvh.py]: fix because of the fix above

File: utils/data.py [hunk 22]
```diff
@@ -93,7 +93,7 @@ def str():
     parents_list = list(map(np.array, parents_list))  # convert list items to np
 
 
-class Edge():
+class Edge:
```
Comment (sigal-raab) [inline utils/data.py]: Erase class Edge

File: utils/data.py [hunk 23]
```diff
@@ -522,30 +465,7 @@ def to_cpu(motion_data):
...
-    anims = [None]*n_motions
-    for idx, one_edge_rot_dict in enumerate(edge_rots_dict):
-        anims[idx], names = anim_from_edge_rot_dict(one_edge_rot_dict)
-    return anims, names
-
-
-def anim_from_edge_rot_dict(edge_rot_dict, root_name='Hips'):
+def basic_anim_from_rot(edge_rot_dict, root_name='Hips'):
     assert root_name in edge_rot_dict['names_with_root']
     root_idx = np.where(edge_rot_dict['names_with_root'] == root_name)[0][0]
```
Comment (sigal-raab) [inline utils/data.py]: when is this code used?

File: utils/data.py [hunk 24]
```diff
@@ -557,11 +477,18 @@ def anim_from_edge_rot_dict(edge_rot_dict, root_name='Hips'):
...
+    if rotations.shape[-1] == 6:  # repr6d
         from Motion.transforms import repr6d2quat
         rotations = repr6d2quat(rotations)
     anim_edges = Animation(rotations, positions, orients, offsets, parents)
 
+    return anim_edges
+
+
+def anim_from_edge_rot_dict(edge_rot_dict, root_name='Hips'):
+    n_frames = edge_rot_dict['rot_edge_no_root'].shape[0]
```
Comment (sigal-raab) [inline utils/data.py]: when is this code used?

File: train_encoder.py [hunk 25]
```diff
@@ -1,27 +1,24 @@
...
 from models.inverse_losses import DiscriminatorLoss, LatentCenterRegularizer, FootContactUnsupervisedLoss
-
-from utils.visualization import motion2bvh
+from utils.visualization import motion2bvh_rot
 from utils.data import calc_bone_lengths, sample_data, requires_grad, data_sampler
 from utils.traits import *
-
 from models.gan import Generator, Discriminator
-
 from utils.data import Joint, Edge # to be used in 'eval'
```
Comment (sigal-raab) [inline train_encoder.py]: remove Edge

File: utils/data.py (hunk 24 above)
Comment (sigal-raab) [reply utils/data.py]: if it's only for preprocess_edges, move it to that file.

File: utils/data.py (hunk 23 above)
Comment (natsala13) [reply utils/data.py]: during sanity check in pre preocess, I will look it up trying to change it to use staticData also.

File: train.py (hunk 17 above)
Comment (natsala13) [reply train.py:200]: I didn't delete them, they appear in the definition of pbar itself a few line earlier.
This feature still exists - before that fix the tqdm prints were not correct.

File: evaluate.py (hunk 6 above)
Comment (natsala13) [reply evaluate.py:60]: I fixed it back.

File: evaluate.py [hunk 26]
```diff
@@ -55,46 +56,54 @@ def generate(args, g_ema, device, mean_joints, std_joints, entity):
...
+    if args.entity == 'Joint':
+        return np.concatenate(generated_motion_np, axis=0)
 
-    _, _, _, edge_rot_dict_general = motion_from_raw(args, np.load(args.path, allow_pickle=True))
-    generated_motions = convert_motions_to_location(args, generated_motion_np, edge_rot_dict_general)
-    return generated_motions
+    motion_data_raw = np.load(args.path, allow_pickle=True)
+    offsets = np.concatenate([motion_data_raw[0]['offset_root'][np.newaxis, :] * 0, motion_data_raw[0]['offsets_no_root']])
+    if args.dataset == 'mixamo':
+        offsets /= 100  # not needed in humanact  TODO: Why do we need that?
```
Comment (sigal-raab) [inline evaluate.py]: remove all TODO in code

File: evaluate.py [hunk 27]
```diff
@@ -55,46 +56,54 @@ def generate(args, g_ema, device, mean_joints, std_joints, entity):
...
-        if args.dataset == 'mixamo':
+        if dataset_type == 'mixamo':
             positions_15joints_oriented = positions_15joints_oriented[:, [0, 2, 1]]
             positions_15joints_oriented[:, 1, :] = -1 * positions_15joints_oriented[:, 1, :]
         generated_motions.append(positions_15joints_oriented)
 
     generated_motions = np.asarray(generated_motions)
     return generated_motions
 
-#endregion
```
Comment (sigal-raab) [inline evaluate.py:97]: why removed #endregion? did you remove the matching #startregion?

File: evaluate.py [hunk 28]
```diff
@@ -225,6 +234,7 @@ def main(args_not_parsed):
     fig_hist_generated = plt.figure()
     plt.bar(*np.unique(yhat.cpu(), return_counts=True))
     plt.title(f'generated {args.ckpt}')
+    plt.savefig('generated_bar_plot.png')
```
Comment (sigal-raab) [inline evaluate.py]: in what path will the figure be saved? use a user-given path or refrain from saving.

File: generate.py [hunk 29]
```diff
@@ -168,21 +156,15 @@ def get_gen_mot_np(args, generated_motion, mean_joints, std_joints):
...
+def generate(args, g_ema: Generator, device, mean_joints: torch.tensor, std_joints: torch.tensor,
+             motion_statics: StaticData, entity: str):
 
     type2func = {'interp': interpolate, 'sample': sample, 'edit': edit}
     with torch.no_grad():
         g_ema.eval()
         mean_latent = g_ema.mean_latent(args.truncation_mean)
         generated_motions = type2func[args.type](args, g_ema, device, mean_latent)
 
-    if entity.str() == 'Joint':
```
Comment (sigal-raab) [inline generate.py:179]: are you sure there is no need for a special care for Joint?

File: models/inverse_losses.py [hunk 30]
```diff
@@ -27,29 +29,27 @@ def forward(self, input):
...
+        motion_data = motion_data * self.normalisation_data['std'][:, :, :motion_data.shape[2]] + \
+                      self.normalisation_data['mean'][:, :, :motion_data.shape[2]]
+        # samples x features x joints x frames  ==>  samples x frames x joints x features
+        motion_for_fk = motion_data.transpose(1, 3)
+
         if self.use_glob_pos:
             #  last 'joint' is global position. use only first 3 features out of it.
             glob_pos = motion_for_fk[:, :, self.pos_offset, :3]
-            if 'use_velocity' in edge_rot_dict_general and edge_rot_dict_general['use_velocity']:
+            if self.normalisation_data['use_velocity']:
```
Comment (sigal-raab) [inline models/inverse_losses.py]: why is user_velocity kept as part of normalization_data?

File: utils/data.py [hunk 31]
```diff
@@ -89,146 +89,7 @@ def str():
                [-1,0],
                [3,0,0,-1,3,3],
                [5,0,1,0,3,-1,5,6,5,8],
-               oj.name2idx(oj.parent_joint_names)]  #full skeleton possesses openpose topology
-    parents_list = list(map(np.array, parents_list))  # convert list items to np
```
Comment (sigal-raab) [inline utils/data.py:93]: no need for class joint's parents_list any more?

File: utils/data.py [hunk 32]
```diff
@@ -738,37 +387,28 @@ def motion_from_raw(args, motion_data_raw):
...
-        edge_rot_dict_general['mean'] = mean_joints.transpose(0, 2, 1, 3)
-        edge_rot_dict_general['std'] = std_joints.transpose(0, 2, 1, 3)
-
-        edge_rot_dict_general['mean_tensor'] = torch.tensor(edge_rot_dict_general['mean'], dtype=torch.float32).cuda()
-        edge_rot_dict_general['std_tensor'] = torch.tensor(edge_rot_dict_general['std'], dtype=torch.float32).cuda()
-
-    # motion_data = motion_data[:1,...]; assert args.batch == 1; print('*********\nOVERFITTING!!!\n******') # overfit over one sample only
-
-    return motion_data, mean_joints, std_joints, edge_rot_dict_general
+    return motion_data, mean_joints, std_joints
```
Comment (sigal-raab) [inline utils/data.py]: why return mean/std if we are using normalization_data?

File: motion_class.py [hunk 33]
```diff
@@ -0,0 +1,604 @@
...
+def plot(name):
+    def decorator(plot_method):
+        def wrapper(self, *args, **kwargs):
+            plt.figure()
+            plot_method(self, *args, **kwargs)
+            plt.title(name)
+            plt.savefig(f'fc/train_{name.replace(" ", "_")}')
+        return wrapper
+    return decorator
```
Comment (sigal-raab) [inline motion_class.py:534]: when is plot() used? 
shouldn't it be in utils/visualization.py?

File: evaluate.py (hunk 27 above)
Comment (natsala13) [reply evaluate.py:97]: I don't see any startregion, also I don't know this syntax, TO ASK.

File: generate.py (hunk 29 above)
Comment (natsala13) [reply generate.py:179]: Yes, this section only prepare the edge_rot_dict var differently wether we're using Joint or Edge entity, since we don't need that variable anymore, we don't need that if as well.
The real separation between Joint and Edge will arrive later on line 208 in the method motion2fig.

File: utils/data.py (hunk 31 above)
Comment (natsala13) [reply utils/data.py:93]: Its the same variable, just a conversion to numpy array which we don't need anymore.

File: motion_class.py (hunk 33 above)
Comment (natsala13) [reply motion_class.py:534]: Its a decorator for all the plot methods of the class DebugDynamic, So its used only for debug.
I added a doc line, but I am not sure this entire class of DebugDynamic is necessary at all.

File: utils/preprocess_edges.py [hunk 34]
```diff
@@ -427,9 +469,9 @@ def preprocess_edge_rot(data_root, out_suffix, dilute_intern_joints, clip_len, s
     ###
     # save final files
     ###
-    anim_edges_split_all_chars_file_name = osp.join(data_root, 'edge_rot'+out_suffix+'.npy')
```
Comment (natsala13) [inline utils/preprocess_edges.py:430]: restore absolute file names.

File: motion_class.py [hunk 35]
```diff
@@ -0,0 +1,604 @@
...
+                anim_exp.rotations[:, parent_idx] = anim_exp.rotations[:, children_one_joint[0]]
+            else:
+                anim_exp.rotations[:, parent_idx] = Quaternions.id((self.n_frames))
+
+        return anim_exp
+
+    def anim_from_static(self):
+        anim_edges = self.basic_anim()
+
+        # TODO: Why do we need to sort names? isnt it already sorted correctly?
```
Comment (natsala13) [inline motion_class.py]: restore sort names

File: motion_class.py [hunk 36]
```diff
@@ -0,0 +1,604 @@
...
+            if len(children_one_joint) > 0:  # not leaf
+                assert len(children_one_joint) == 1 or (anim_exp.offsets[children_one_joint] == np.zeros(3)).all() and (
+                        anim_exp.rotations[:, children_one_joint] == Quaternions.id((self.n_frames, 1))).all()
+                anim_exp.rotations[:, parent_idx] = anim_exp.rotations[:, children_one_joint[0]]
+            else:
+                anim_exp.rotations[:, parent_idx] = Quaternions.id((self.n_frames))
+
+        return anim_exp
+
+    def anim_from_static(self):
```
Comment (natsala13) [inline motion_class.py]: change name - maybe to_anim

File: models/inverse_losses.py (hunk 30 above)
Comment (natsala13) [reply models/inverse_losses.py]: because it is loaded in second time from the checkpoint and it is not forworded back to the rest of the params.

Comment (natsala13) [reply models/inverse_losses.py]: Changed
